network/kademlia_load_balancer_test.go

```diff
@@ -0,0 +1,473 @@
+	useStats.WaitKey(bitStringToHex("10010100"))
+	tk.On("10010001") // bin 1 dec 145
+	useStats.WaitKey(bitStringToHex("10010001"))
+	tk.On("11010100") // bin 2 dec 212
+	useStats.WaitKey(bitStringToHex("11010100"))
+	tk.On("11010101") // bin 2 dec 213
+	useStats.WaitKey(bitStringToHex("11010101"))
+	stats := make(map[string]int)
+	countUse := func(bin LBBin) bool {
+		peer := bin.LBPeers[0].Peer
+		bin.LBPeers[0].AddUseCount()
+		key := peerToBitString(peer)
+		stats[key] = stats[key] + 1
+		return false
+	}
+
+	pivotAddressBin1 := pot.NewAddressFromString("10000000") // Two nearest peers (1,2)
+	// Instead of selecting peers 10010100 or 10010001, capPeer is always chosen (10100000)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+
+	count := useStats.GetUses(capPeer)
+	if count != 3 || stats["10100000"] != 3 {
+		t.Errorf("Expected 3 uses of capability peer but got %v/%v", count, stats["10100000"])
+	}
+
+	secondCapPeer := tk.newTestKadPeerWithCapabilities("10100001", caps[capKey])
+	tk.Kademlia.On(secondCapPeer)
+	useStats.WaitKey(secondCapPeer.Key())
+	secondCountStart := useStats.GetUses(secondCapPeer)
+	count = useStats.GetUses(capPeer)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	klb.EachBinFiltered(pivotAddressBin1, capKey, countUse)
+	secondCount := useStats.GetUses(secondCapPeer)
+	if secondCount-secondCountStart != 2 {
+		t.Errorf("Expected 2 uses of second capability peer but got %v", secondCount-secondCountStart)
+	}
+
+}
+
+type testKademliaBackend struct {
+	baseAddr       []byte
+	addedChannel   *pubsubchannel.PubSubChannel
+	removedChannel *pubsubchannel.PubSubChannel
+	bins           map[int][]*Peer
+	maxPo          int
+}
+
+// EachConn iterates this test kademlia table peers in order po from nearest to furthest with respect to the base address.
+// - First it takes the po of the base address provided. With this po it takes the corresponding bin B (peers with same po
+// as this base) copy the list and sort them using sort.Slice from furthest to lowest. Then it calls the provided consume
+// function with those peers and its corresponding po (peerPo).
+// - Then, it iterates all bins with higher po than B, and iterates all peers. All of these peers will have the
+// same po as the base address with respect to the pin of the kademlia, that's why we don't need to sort them out and
+// we can iterate them in any order. For each peer the consume function is called with the same po.
+// - Finally, the bins furthest than B are iterated from highest po (nearest) to lowest (furthest). For each bin, all
+// peers are iterated and called the consume function with the peer and the bin po.
+// After any of the calls to the consume function, if that function returns false, the iteration stops.
+func (tkb *testKademliaBackend) EachConn(base []byte, maxPo int, consume func(*Peer, int) bool) {
```

Comment:
can you justify why this mock is needed any why you need to mock `pot` and `kademlia` functionality instead of using them directly?
needless to say that if some of the code in `kademlia` and `pot` changes (and it would take only the slightest change) - none of the tests using the mock will fail

Reply:
Well, that will be and integration test, not a unit test, but maybe I can use newTestKademlia there without much change.

Reply:
I have removed testKademliaBackend, but his will make harder to move kademlia_load_balancer to its own package as now depends even more of network test classes.